### PR TITLE
Separate peer discovery from transport

### PR DESCRIPTION
**Pull request by limebell**
This patch separates peer discovery logic from transport.

Skipped changelog for `NetMQTransport` constructor because it is added in 0.11.

### CODE REVIEW

File: Libplanet/Net/Protocols/IProtocol.cs [hunk 1]
```diff
@@ -1,30 +1,85 @@
...
+        /// that this operation should be canceled.</param>
+        /// <returns>An awaitable task without value.</returns>
         Task RefreshTableAsync(TimeSpan maxAge, CancellationToken cancellationToken);
 
+        /// <summary>
+        /// Reconstructs network connection between peers on network.
+        /// </summary>
+        /// <param name="cancellationToken">A cancellation token used to propagate notification
+        /// that this operation should be canceled.</param>
+        /// <returns>>An awaitable task without value.</returns>
```
Comment (longfin) [inline Libplanet/Net/Protocols/IProtocol.cs:72]:
```suggestion
        /// <returns>An awaitable task without value.</returns>
```

File: Libplanet/Net/Protocols/IProtocol.cs [hunk 2]
```diff
@@ -1,30 +1,85 @@
...
         Task RebuildConnectionAsync(CancellationToken cancellationToken);
 
+        /// <summary>
+        /// Checks the <see cref="KBucket"/> in the <see cref="RoutingTable"/> and if
+        /// there is an empty <see cref="KBucket"/>, fill it with <see cref="Peer"/>s
+        /// in the <see cref="KBucket.ReplacementCache"/>.
+        /// </summary>
+        /// <param name="cancellationToken">A cancellation token used to propagate notification
+        /// that this operation should be canceled.</param>
+        /// <returns>>An awaitable task without value.</returns>
```
Comment (longfin) [inline Libplanet/Net/Protocols/IProtocol.cs:82]:
```suggestion
        /// <returns>An awaitable task without value.</returns>
```

File: Libplanet/Net/Protocols/KademliaProtocol.cs [hunk 3]
```diff
@@ -18,48 +18,38 @@ internal class KademliaProtocol : IProtocol
...
             _random = new System.Random();
-            _tableSize = tableSize ?? Kademlia.TableSize;
-            _bucketSize = bucketSize ?? Kademlia.BucketSize;
             _findConcurrency = findConcurrency;
-            _routing = new RoutingTable(_address, _tableSize, _bucketSize, _random, _logger);
+            _table = table;
             _requestTimeout =
                 requestTimeout ??
-                TimeSpan.FromMilliseconds(Kademlia.IdleRequestTimeout);
+                TimeSpan.FromMilliseconds(5000);
```
Comment (longfin) [inline Libplanet/Net/Protocols/KademliaProtocol.cs:42]: Is it okay we use `5000` directly instead of named constant?

**Review by longfin — APPROVED** [review]

File: Libplanet/Net/Protocols/KademliaProtocol.cs (hunk 3 above)
Comment (limebell) [reply Libplanet/Net/Protocols/KademliaProtocol.cs:42]: I thought it is weird that timeout value is located at Kademlia constants.

File: Libplanet/Net/Protocols/RoutingTable.cs [hunk 4]
```diff
@@ -34,23 +40,47 @@ internal class RoutingTable
...
+        /// </summary>
+        public int BucketSize { get; }
+
+        /// <summary>
+        /// Number of peers in the table.
+        /// </summary>
         public int Count => _buckets.Sum(bucket => bucket.Count);
 
+        /// <summary>
+        /// <see cref="IEnumerable{T}"/> of peers in the table.
```
Comment (moreal) [inline Libplanet/Net/Protocols/RoutingTable.cs:71]:
```suggestion
        /// List of peers in the table.
```

With the XML document, we will see a strange sentence like the below link.

image


https://docs.libplanet.io/main/api/Libplanet.Net.ITransport.html#Libplanet_Net_ITransport_BootstrapAsync_IEnumerable_Libplanet_Net_BoundPeer__System_Nullable_TimeSpan__System_Nullable_TimeSpan__System_Int32_CancellationToken_

**Review by moreal — COMMENTED** [review]
It seems to need to resolve conflict 🙏

File: Libplanet/Net/Protocols/RoutingTable.cs (hunk 4 above)
Comment (dahlia) [reply Libplanet/Net/Protocols/RoutingTable.cs:71]: @moreal That's a known issue (at least to me), and it's the problem with our build settings with [DocFX].  We shouldn't remove the `<see>` element due to this, but we should fix the DocFX configuration instead.

[DocFX]: https://dotnet.github.io/docfx/

**Review by dahlia — DISMISSED** [review]

**Review by longfin — APPROVED** [review]

**Review by moreal — APPROVED** [review]

**Review by longfin — DISMISSED** [review]

**Review by moreal — APPROVED** [review]

**Review by dahlia — COMMENTED** [review]
This should be rebased again.

**Comment by riemannulus** [comment]
Seems conflict 😢

**Review by moreal — DISMISSED** [review]

**Review by dahlia — APPROVED** [review]

**Review by longfin — APPROVED** [review]

**Review by longfin — APPROVED** [review]

**Review by moreal — APPROVED** [review]

**Pull request merged**